### PR TITLE
BigQueryDataSourceV2 writes BigNumeric as BigNumeric to BQ

### PR DESCRIPTION
**Pull request by himanshukohli09**

### CODE REVIEW

File: connector/src/main/java/com/google/cloud/spark/bigquery/SchemaConverters.java [hunk 1]
```diff
@@ -15,19 +15,12 @@
...
 
-import com.google.cloud.bigquery.Field;
-import com.google.cloud.bigquery.FieldList;
-import com.google.cloud.bigquery.LegacySQLTypeName;
-import com.google.cloud.bigquery.Schema;
-import com.google.cloud.bigquery.StandardTableDefinition;
-import com.google.cloud.bigquery.TableDefinition;
-import com.google.cloud.bigquery.TableInfo;
-import com.google.cloud.bigquery.TimePartitioning;
+import com.google.cloud.bigquery.*;
```
Comment (davidrabinowitz) [inline connector/src/main/java/com/google/cloud/spark/bigquery/SchemaConverters.java:18]: Please avoid import `package.*`, import the actual classes

File: connector/src/test/java/com/google/cloud/spark/bigquery/AvroSchemaConverterTest.java [hunk 2]
```diff
@@ -32,6 +32,8 @@
 import java.math.BigDecimal;
 import java.nio.ByteBuffer;
 
+import static com.google.cloud.spark.bigquery.SchemaConverters.*;
```
Comment (davidrabinowitz) [inline connector/src/test/java/com/google/cloud/spark/bigquery/AvroSchemaConverterTest.java:35]: Please avoid `import static Class.*`, import the actual fields/methods

File: connector/third_party/apache-spark/src/main/java/org/apache/spark/bigquery/BigNumeric.java [hunk 3]
```diff
@@ -19,4 +19,17 @@ public BigDecimal getNumber() {
     public String toString() {
         return this.number.toPlainString();
     }
+
+    @Override
+    public boolean equals(Object other) {
```
Comment (davidrabinowitz) [inline connector/third_party/apache-spark/src/main/java/org/apache/spark/bigquery/BigNumeric.java:24]: Missing hascode()

**Comment by davidrabinowitz** [comment]
`BigNumeric` Should be serializable:
```
ava.io.NotSerializableException: org.apache.spark.bigquery.BigNumeric
Serialization stack:
	- object not serializable (class: org.apache.spark.bigquery.BigNumeric, value: 1.0)
	- element of array (index: 0)
	- array (class [Ljava.lang.Object;, size 1)
	- field (class: org.apache.spark.sql.catalyst.expressions.GenericRow, name: values, type: class [Ljava.lang.Object;)
	- object (class org.apache.spark.sql.catalyst.expressions.GenericRow, [1.0])
	- element of array (index: 0)
	- array (class [Lorg.apache.spark.sql.Row;, size 1)
	- field (class: scala.collection.mutable.WrappedArray$ofRef, name: array, type: class [Ljava.lang.Object;)
	- object (class scala.collection.mutable.WrappedArray$ofRef, WrappedArray([1.0]))
	- writeObject data (class: org.apache.spark.rdd.ParallelCollectionPartition)
	- object (class org.apache.spark.rdd.ParallelCollectionPartition, org.apache.spark.rdd.ParallelCollectionPartition@698)
	- field (class: org.apache.spark.scheduler.ResultTask, name: partition, type: interface org.apache.spark.Partition)
	- object (class org.apache.spark.scheduler.ResultTask, ResultTask(2, 7))
	at org.apache.spark.serializer.SerializationDebugger$.improveException(SerializationDebugger.scala:41)
	at org.apache.spark.serializer.JavaSerializationStream.writeObject(JavaSerializer.scala:47)
	at org.apache.spark.serializer.JavaSerializerInstance.serialize(JavaSerializer.scala:101)
	at org.apache.spark.scheduler.TaskSetManager.$anonfun$resourceOffer$2(TaskSetManager.scala:460)
	at scala.Option.map(Option.scala:230)
	at org.apache.spark.scheduler.TaskSetManager.resourceOffer(TaskSetManager.scala:444)
	at org.apache.spark.scheduler.TaskSchedulerImpl.$anonfun$resourceOfferSingleTaskSet$2(TaskSchedulerImpl.scala:397)
	at org.apache.spark.scheduler.TaskSchedulerImpl.$anonfun$resourceOfferSingleTaskSet$2$adapted(TaskSchedulerImpl.scala:392)
	at scala.Option.foreach(Option.scala:407)
	at org.apache.spark.scheduler.TaskSchedulerImpl.$anonfun$resourceOfferSingleTaskSet$1(TaskSchedulerImpl.scala:392)
	at scala.collection.immutable.Range.foreach$mVc$sp(Range.scala:158)
	at org.apache.spark.scheduler.TaskSchedulerImpl.resourceOfferSingleTaskSet(TaskSchedulerImpl.scala:383)
	at org.apache.spark.scheduler.TaskSchedulerImpl.$anonfun$resourceOffers$20(TaskSchedulerImpl.scala:581)
	at org.apache.spark.scheduler.TaskSchedulerImpl.$anonfun$resourceOffers$20$adapted(TaskSchedulerImpl.scala:576)
	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:198)
	at org.apache.spark.scheduler.TaskSchedulerImpl.$anonfun$resourceOffers$16(TaskSchedulerImpl.scala:576)
	at org.apache.spark.scheduler.TaskSchedulerImpl.$anonfun$resourceOffers$16$adapted(TaskSchedulerImpl.scala:547)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at org.apache.spark.scheduler.TaskSchedulerImpl.resourceOffers(TaskSchedulerImpl.scala:547)
	at org.apache.spark.scheduler.local.LocalEndpoint.reviveOffers(LocalSchedulerBackend.scala:91)
	at org.apache.spark.scheduler.local.LocalEndpoint$$anonfun$receive$1.applyOrElse(LocalSchedulerBackend.scala:68)
	at org.apache.spark.rpc.netty.Inbox.$anonfun$process$1(Inbox.scala:115)
	at org.apache.spark.rpc.netty.Inbox.safelyCall(Inbox.scala:213)
	at org.apache.spark.rpc.netty.Inbox.process(Inbox.scala:100)
	at org.apache.spark.rpc.netty.MessageLoop.org$apache$spark$rpc$netty$MessageLoop$$receiveLoop(MessageLoop.scala:75)
	at org.apache.spark.rpc.netty.MessageLoop$$anon$1.run(MessageLoop.scala:41)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

**Comment by himanshukohli09** [comment]
> `BigNumeric` Should be serializable:
> 
> ```
> ava.io.NotSerializableException: org.apache.spark.bigquery.BigNumeric
> Serialization stack:
> 	- object not serializable (class: org.apache.spark.bigquery.BigNumeric, value: 1.0)
> 	- element of array (index: 0)
> 	- array (class [Ljava.lang.Object;, size 1)
> 	- field (class: org.apache.spark.sql.catalyst.expressions.GenericRow, name: values, type: class [Ljava.lang.Object;)
> 	- object (class org.apache.spark.sql.catalyst.expressions.GenericRow, [1.0])
> 	- element of array (index: 0)
> 	- array (class [Lorg.apache.spark.sql.Row;, size 1)
> 	- field (class: scala.collection.mutable.WrappedArray$ofRef, name: array, type: class [Ljava.lang.Object;)
> 	- object (class scala.collection.mutable.WrappedArray$ofRef, WrappedArray([1.0]))
> 	- writeObject data (class: org.apache.spark.rdd.ParallelCollectionPartition)
> 	- object (class org.apache.spark.rdd.ParallelCollectionPartition, org.apache.spark.rdd.ParallelCollectionPartition@698)
> 	- field (class: org.apache.spark.scheduler.ResultTask, name: partition, type: interface org.apache.spark.Partition)
> 	- object (class org.apache.spark.scheduler.ResultTask, ResultTask(2, 7))
> 	at org.apache.spark.serializer.SerializationDebugger$.improveException(SerializationDebugger.scala:41)
> 	at org.apache.spark.serializer.JavaSerializationStream.writeObject(JavaSerializer.scala:47)
> 	at org.apache.spark.serializer.JavaSerializerInstance.serialize(JavaSerializer.scala:101)
> 	at org.apache.spark.scheduler.TaskSetManager.$anonfun$resourceOffer$2(TaskSetManager.scala:460)
> 	at scala.Option.map(Option.scala:230)
> 	at org.apache.spark.scheduler.TaskSetManager.resourceOffer(TaskSetManager.scala:444)
> 	at org.apache.spark.scheduler.TaskSchedulerImpl.$anonfun$resourceOfferSingleTaskSet$2(TaskSchedulerImpl.scala:397)
> 	at org.apache.spark.scheduler.TaskSchedulerImpl.$anonfun$resourceOfferSingleTaskSet$2$adapted(TaskSchedulerImpl.scala:392)
> 	at scala.Option.foreach(Option.scala:407)
> 	at org.apache.spark.scheduler.TaskSchedulerImpl.$anonfun$resourceOfferSingleTaskSet$1(TaskSchedulerImpl.scala:392)
> 	at scala.collection.immutable.Range.foreach$mVc$sp(Range.scala:158)
> 	at org.apache.spark.scheduler.TaskSchedulerImpl.resourceOfferSingleTaskSet(TaskSchedulerImpl.scala:383)
> 	at org.apache.spark.scheduler.TaskSchedulerImpl.$anonfun$resourceOffers$20(TaskSchedulerImpl.scala:581)
> 	at org.apache.spark.scheduler.TaskSchedulerImpl.$anonfun$resourceOffers$20$adapted(TaskSchedulerImpl.scala:576)
> 	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
> 	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
> 	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:198)
> 	at org.apache.spark.scheduler.TaskSchedulerImpl.$anonfun$resourceOffers$16(TaskSchedulerImpl.scala:576)
> 	at org.apache.spark.scheduler.TaskSchedulerImpl.$anonfun$resourceOffers$16$adapted(TaskSchedulerImpl.scala:547)
> 	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
> 	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
> 	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
> 	at org.apache.spark.scheduler.TaskSchedulerImpl.resourceOffers(TaskSchedulerImpl.scala:547)
> 	at org.apache.spark.scheduler.local.LocalEndpoint.reviveOffers(LocalSchedulerBackend.scala:91)
> 	at org.apache.spark.scheduler.local.LocalEndpoint$$anonfun$receive$1.applyOrElse(LocalSchedulerBackend.scala:68)
> 	at org.apache.spark.rpc.netty.Inbox.$anonfun$process$1(Inbox.scala:115)
> 	at org.apache.spark.rpc.netty.Inbox.safelyCall(Inbox.scala:213)
> 	at org.apache.spark.rpc.netty.Inbox.process(Inbox.scala:100)
> 	at org.apache.spark.rpc.netty.MessageLoop.org$apache$spark$rpc$netty$MessageLoop$$receiveLoop(MessageLoop.scala:75)
> 	at org.apache.spark.rpc.netty.MessageLoop$$anon$1.run(MessageLoop.scala:41)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
> 	at java.base/java.lang.Thread.run(Thread.java:829)
> ```

Done

File: connector/src/main/java/com/google/cloud/spark/bigquery/SchemaConverters.java (hunk 1 above)
Comment (himanshukohli09) [reply connector/src/main/java/com/google/cloud/spark/bigquery/SchemaConverters.java]: Done

File: connector/third_party/apache-spark/src/main/java/org/apache/spark/bigquery/BigNumeric.java (hunk 3 above)
Comment (himanshukohli09) [reply connector/third_party/apache-spark/src/main/java/org/apache/spark/bigquery/BigNumeric.java:26]: Done

File: connector/src/test/java/com/google/cloud/spark/bigquery/AvroSchemaConverterTest.java (hunk 2 above)
Comment (himanshukohli09) [reply connector/src/test/java/com/google/cloud/spark/bigquery/AvroSchemaConverterTest.java]: Done

File: connector/third_party/apache-spark/src/main/java/org/apache/spark/bigquery/BigNumeric.java [hunk 4]
```diff
@@ -1,10 +1,12 @@
 package org.apache.spark.bigquery;
 
 import org.apache.spark.sql.types.SQLUserDefinedType;
+
+import java.io.Serializable;
 import java.math.BigDecimal;
 
 @SQLUserDefinedType(udt = BigNumericUDT.class)
-public class BigNumeric {
+public class BigNumeric implements Serializable {
```
Comment (davidrabinowitz) [inline connector/third_party/apache-spark/src/main/java/org/apache/spark/bigquery/BigNumeric.java:9]: Missing [serialVersionUID](https://www.baeldung.com/java-serial-version-uid)

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: connector/src/test/java/com/google/cloud/spark/bigquery/SupportedCustomDataTypeTest.java [hunk 5]
```diff
@@ -37,4 +42,30 @@ public void testMatrix() throws Exception {
...
+  }
+
+  @Test
+  public void testBigNumericHashCode() throws IOException {
+    BigNumeric bn1 = new BigNumeric(new BigDecimal("123.123"));
+    BigNumeric bn2 = new BigNumeric(new BigDecimal("123.123"));
+    BigNumeric bn3 = new BigNumeric(new BigDecimal("456.456"));
+
+    assertThat(bn1.hashCode() == bn2.hashCode()).isTrue();
+    assertThat(bn1.hashCode() == bn3.hashCode()).isFalse();
```
[the remark this answers is not in the document]
Comment (himanshukohli09) [reply connector/src/test/java/com/google/cloud/spark/bigquery/SupportedCustomDataTypeTest.java]: done

File: connector/src/test/java/com/google/cloud/spark/bigquery/SupportedCustomDataTypeTest.java [hunk 6]
```diff
@@ -37,4 +42,30 @@ public void testMatrix() throws Exception {
...
+        .writeObject(new BigNumeric(new BigDecimal("123.123")));
+  }
+
+  @Test
+  public void testBigNumericEquality() throws IOException {
+    BigNumeric bn1 = new BigNumeric(new BigDecimal("123.123"));
+    BigNumeric bn2 = new BigNumeric(new BigDecimal("123.123"));
+    BigNumeric bn3 = new BigNumeric(new BigDecimal("456.456"));
+
+    assertThat(bn1.equals(bn2)).isTrue();
```
[the remark this answers is not in the document]
Comment (himanshukohli09) [reply connector/src/test/java/com/google/cloud/spark/bigquery/SupportedCustomDataTypeTest.java]: done

File: connector/third_party/apache-spark/src/main/java/org/apache/spark/bigquery/BigNumeric.java (hunk 4 above)
Comment (himanshukohli09) [reply connector/third_party/apache-spark/src/main/java/org/apache/spark/bigquery/BigNumeric.java:9]: done

**Review by davidrabinowitz — APPROVED** [review]

**Pull request merged**